### ISSUE TITLE
`Behavior a -> Event b -> Dynamic a` please?

### ISSUE THREAD

**Issue by lspitzner** [reflex-frp/reflex#140]
Really i was looking for `(a -> b -> c) -> Behavior a -> Dynamic b -> Dynamic c`. Both are expressible via some combination of `updated`, `attachWith`, `current`, `sample` and `buildDynamic`, but i'd prefer something a bit higher-level.

(i base this mostly on the quickref, so perhaps i am missing something that isn't there.)

**Comment by 3noch** [comment]
I've wanted this before as well.

**Comment by ryantrinkle** [comment]
@lspitzner What are the semantics you're looking for?  E.g.: if we produce `out` using this, when should `updated out` fire, and what should be the value of `current out` at all times?

**Comment by 3noch** [comment]
I would imagine
`mkDynamic (current x) (updated x) == x`.

**Comment by ryantrinkle** [comment]
I thought about these some more, and I think these functions can't be written.

In the case of `f :: (a -> b -> c) -> Behavior a -> Dynamic b -> Dynamic c`, suppose we do: `dab = f (,) ba (pure b)`.  The only `Event` in sight is `updated (pure b)`, which is equivalent to `never`.  Therefore, the `updated dab` cannot fire, since we can't spontaneously generate `Event` occurrences in pure code.  The "Dynamic law" is that `forall d, current d changes when and only when updated d fires`, so `current dab` cannot ever change.  This means that `dab = pure ab` for some `ab = (a, b) :: (a, b)`.  However, the only way we can get an `a` is from `ba`, and this is a pure function, which means it must operate identically regardless of when it is evaluated.  As a result, it cannot choose any particular time at which to sample `ba`.  Since `ba` may not be constant, `a` is ill-defined.

So, I believe that the only acceptable implementation for this function involves supplying `undefined` for `a` at least some (I would guess _all_, actually) of the time.

Note that this argument makes critical use of the Dynamic law, so, as you might expect, `unsafeBuildDynamic` allows you to get around this.  This will only work properly if there is some special relationship between `ba` and the `Dynamic b` that ensures that `dab` will be lawful - and these things get *really* subtle.


For `mkDynamic`, the proof is simpler: in `mkDynamic (current x) (updated x) == x'`, the only `a` we have is from `current x :: Behavior a`, since the second argument has type `Event b` (and `a` is not generally the same type as `b`).  Since `Behaviors` are never prompt, we cannot produce `updated x' == updated x`.

**Comment by 3noch** [comment]
What if `mkDynamic :: Behavior a -> Event a -> Dynamic a`?

**Comment by ryantrinkle** [comment]
The function with that exact type signature is [`unsafeDynamic`](https://github.com/reflex-frp/reflex/blob/36db894180d015fdd032d7d3a338e0b5ac9df4b4/src/Reflex/Class.hs#L426).  It's unsafe because it has no way to know whether your `Event` and `Behavior` obey the Dynamic law.

[`buildDynamic`](https://github.com/reflex-frp/reflex/blob/60d2878142943487987feeeea108dbed5405f469/src/Reflex/Class.hs#L465), on the other hand, is monadic, and I think it's got almost exactly the semantics you're describing.  If you write `d <-
 buildDynamic (sample b) e`, you get a `Dynamic` whose value is the same as `b` when it's created, and changes when (and only when) `e` fires.  It's the same semantics as `v0 <- sample b ; holdDyn v0 e`, except that it's lazier (with `buildDynamic`, the `sample b` is forced either when `current d` is forced or before the next frame, whichever comes first - whereas with the other approach, it's forced immediately).

**Comment by 3noch** [comment]
Aha! I did notice that `unsafeDynamic` fit the bill but I didn't want to use it for the obvious reason. But using `buildDynamic` seems like a great way as well!

**Comment by ryantrinkle** [comment]
FWIW, I've found that `unsafeDynamic` is surprisingly difficult to use properly.  I wouldn't recommend it if you have any other options.

**Comment by 3noch** [comment]
It's aptly named. I don't think I'd ever dare to use it.

**Comment by lspitzner** [comment]
@ryantrinkle Right.

What about the versions that return monadic values? Are the following bad idea? (The names may be chosen badly, and I dislike the RankNTypes, but otherwise?)

~~~~.hs
dynamicPushAlways1
  :: (Reflex t, MonadHold t m)
  => (forall m' . MonadSample t m => a -> m' b)
  -> R.Dynamic t a
  -> m (R.Dynamic t b)
dynamicPushAlways1 f ad =
  buildDynamic (f =<< sample (current ad)) (pushAlways f (updated ad))

dynamicPushAlways2
  :: (Reflex t, MonadHold t m, MonadFix m)
  => (forall m' . MonadSample t m => a -> m' b)
  -> Dynamic t a
  -> m (Dynamic t b)
dynamicPushAlways2 f ad = do
  b0 <- f =<< sample (current ad)
  foldDynM (\a _ -> f a) b0 (updated ad)

dynamicAttachWith
  :: (Reflex t, MonadSample t m, MonadHold t m, MonadFix m)
  => (a -> b -> c)
  -> Behavior t a
  -> Dynamic t b
  -> m (Dynamic t c)
dynamicAttachWith f ab bd = do
  a0 <- sample ab
  b0 <- sample (current bd)
  holdDyn (f a0 b0) (pushAlways (\b -> sample ab <&> \a -> f a b) (updated bd))

dynamicTag
  :: (Reflex t, MonadSample t m, MonadHold t m, MonadFix m)
  => Behavior t a
  -> Dynamic t b
  -> m (Dynamic t a)
dynamicTag ab bd = do
  a0 <- sample ab
  holdDyn a0 (pushAlways (\_ -> sample ab) (updated bd))
~~~~

**Comment by ryantrinkle** [comment]
@lspitzner Can you provide some examples of the situations where these would be used?

**Comment by lspitzner** [comment]
No, not really :/ Most of the time one can invent some initial values so creating a Dynamic from some pushAlways'd Event is trivial. I'll have to look out for some good examples.

**Comment by ryantrinkle** [comment]
My general sense is that, if I found myself needing `dynamicPushAlways` and `dynamicAttachWith`, I'd wonder whether I had made a mistake in some of the surrounding code.  That's not to say that they're wrong, bad, or anything like that, just that it's tough to imagine when I'd need them.

`dynamicTag` is a bit more obvious, in that it seems to just sample the given behavior at creation and then whenever the given event fires.  I can't think of a use case off the top of my head, but I suspect there are some.  In this case, the question is just: can we think of a name that is shorter and more understandable than the implementation itself, and document it well?  Having concrete use case examples would help with that, I think.  Here's a better (lazier) implementation, by the way:

```
dynamicTag b e = buildDynamic (sample b) $ tag b e
```
